### PR TITLE
Add dummy app component/test, use rootElement to select

### PR DESCRIPTION
**Pull request by pixelhandler**
Add find helper and use within test helpers

- Support same params as Ember's `find` helper, only not using `$`
- `find` helper uses `document.querySelectorAll` and when only one element is found, simple return that element; otherwise the `NodeList`
- Use app config/enviroment setting for `rootElement` (default is `#ember-testing`) - to query within the test DOM (vs all DOM including test runner HTML)

In order to support a `find` (DOM) query only inside the Ember application's DOM the consuming application's config is used. For now in `tests/test-helper.js` the default value for `#ember-testing` as the `rootElement` can be set per the app's config:

```js
import settings from 'ember-native-dom-helpers/test-support/settings';
import config from '../config/environment';
const { APP: { rootElement } } = config;

settings.rootElement = rootElement || settings.rootElement;
```

Also, since these are test helpers I thought it would be a good idea to include an integration test using them. I added a form component in the dummy app and an integration test that uses these helpers.

The `find` helper acts like it uses jQuery's selector but does not so helper are some affordances:

- If a HTMLElement is passed to find, simply return it
- If a NodeList is passed to find, return it
- If selector only matches one HTMLElement return it
- If selector matches more then one element return the NodeList

For the most part the find helper behaves like the find helper for acceptance tests, except is not using jQuery's selector engine, nor returns a wrapped element or set. Instead native DOM element(s) will be returned. Hence the native `find` helper returns one or a set of elements.

In addition to a `find` helper there is a `first` helper as well. The implementations of the acceptance test helpers only operate on the first item found from a DOM query. So the these test helpers use the `first` helper to select the first matching element.

For test suites that have heavy use of jQuery - a setting affords using `find` which can return a jQuery wrapped set (instead of HTMLElement or NodeList of HTMLElements).

```js
import settings from 'ember-native-dom-helpers/test-support/settings';
settings.useJQueryWrapper = true;
```

### CODE REVIEW

File: addon-test-support/helpers.js [hunk 1]
```diff
@@ -7,6 +7,9 @@ const DEFAULT_EVENT_OPTIONS = { canBubble: true, cancelable: true };
 const KEYBOARD_EVENT_TYPES = ['keydown', 'keypress', 'keyup'];
 const MOUSE_EVENT_TYPES = ['click', 'mousedown', 'mouseup', 'dblclick', 'mouseenter', 'mouseleave', 'mousemove', 'mouseout', 'mouseover'];
 
+import config from 'dummy/config/environment';
```
Comment (pixelhandler) [inline addon-test-support/helpers.js]: @rwjblue I had trouble importing the config from the consuming app, using the full path to the `dummy` config worked, but it's not right.

**Comment by pixelhandler** [comment]
@rwjblue @cibernox this PR addresses a bug I found when writing a component test, the `selector` argument passed to the helper method didn't have any awareness of the Ember root element. So by accident I began to `fillIn` the inputs that some with the QUnit test runner, e.g. `filter` instead of my input within the context of the `#ember-testing` element.

Do you think we can utilize the consuming app's rootElement env setting in the helpers?

File: addon-test-support/helpers.js (hunk 1 above)
Comment (rwjblue) [inline addon-test-support/helpers.js]: Right, I don't think we can do this (at least not reliably). We will need to think of another way...

File: addon-test-support/helpers.js [hunk 2]
```diff
@@ -111,6 +114,7 @@ function buildKeyboardEvent(type, options = {}) {
 }
 
 export function click(selector, options = {}) {
+  selector = `${rootElement} ${selector}`;
```
Comment (rwjblue) [inline addon-test-support/helpers.js]: We should generally prefer to avoid mutating an argument name like this (it makes debugging / stepping in chrome difficult).  I'd suggest something like this instead:

```js
export function click(_selector, options = {}) {
  let selector = `${rootElement} ${_selector}`;
}
```

File: addon-test-support/helpers.js [hunk 3]
```diff
@@ -120,6 +124,7 @@ export function click(selector, options = {}) {
 }
 
 export function fillIn(selector, text) {
+  selector = `${rootElement} ${selector}`;
```
Comment (rwjblue) [inline addon-test-support/helpers.js]: Same as above

File: addon-test-support/helpers.js [hunk 4]
```diff
@@ -129,6 +134,7 @@ export function fillIn(selector, text) {
 }
 
 export function triggerEvent(selector, type, options) {
+  selector = `${rootElement} ${selector}`;
```
Comment (rwjblue) [inline addon-test-support/helpers.js]: Same as above

File: ember-cli-build.js [hunk 5]
```diff
@@ -3,7 +3,9 @@ const EmberAddon = require('ember-cli/lib/broccoli/ember-addon');
 
 module.exports = function(defaults) {
   var app = new EmberAddon(defaults, {
-    // Add options here
+    babel: {
+      includePolyfill: true
```
Comment (rwjblue) [inline ember-cli-build.js]: In order to ensure that this library works without the babel polyfill, I think we should not add it...

File: .eslintrc.js [hunk 6]
```diff
@@ -1,5 +1,6 @@
 module.exports = {
   root: true,
+  parser: 'babel-eslint',
```
Comment (rwjblue) [inline .eslintrc.js]: Lets roll this back, and just rely on promise chaining.  I know that is annoying, but its pretty important to ensure that this library doesn't *require* babel-polyfill and whatnot...

File: tests/integration/components/x-form-test.js [hunk 7]
```diff
@@ -0,0 +1,95 @@
+import { moduleForComponent, test } from 'ember-qunit';
+import hbs from 'htmlbars-inline-precompile';
+import { fillIn, click, keyEvent, triggerEvent } from 'ember-native-dom-helpers/test-support/helpers';
+
+moduleForComponent('x-form', 'Integration | Component | x-form', {
+  integration: true
+});
+
+const find = (selector) => document.querySelectorAll(`#ember-testing ${selector}`);
```
Comment (rwjblue) [inline tests/integration/components/x-form-test.js]: We should make a helper in addon-test-support for this ;)

**Comment by cibernox** [comment]
@pixelhandler Let me see if I understand this. You want to targets something like `click('a')` and since there is some link in the qunit/mocha interface, the helper is targeting that?

If I'm right and this is something we care to support (I think we should) I'd create a `find` helper, similar to the one in ember itself: https://github.com/emberjs/ember.js/blob/master/packages/ember-testing/lib/helpers/find.js#L30-L35


And wherever I was using `document.querySelector` we replace it with that. 
Do we want to export it?

**Comment by pixelhandler** [comment]
@cibernox A couple replies…

> You want to targets something like click('a') and since there is some link in the qunit/mocha interface, the helper is targeting that?

Specifically there are `inputs` elements in the QUnit interface and the test context is usually `#ember-testing`; however an app's config/environment file may customize the `rootElement` value.

And yes, we should make a `find` helper that uses the context of the test DOM, for example:

`const find = (selector) => document.querySelector('${rootElement} selector');`

File: ember-cli-build.js (hunk 5 above)
Comment (pixelhandler) [inline ember-cli-build.js]: @rwjblue I added this to support async/await in the tests but yeah can remove.

**Comment by pixelhandler** [comment]
@cibernox @rwjblue I added another commit with a `find` helper

**Comment by pixelhandler** [comment]
I'll rebase this on the branch for #4 since that has a working build for Travis now

**Comment by pixelhandler** [comment]
@cibernox @rwjblue I updated the `find` helper to act like the acceptance test helper which uses jQuery's selector. But without using jQuery:

- If a HTMLElement is passed to find, simply return it
- If a NodeList is passed to find, return it
- If selector only matches one HTMLElement return it
- If selecotor matches more then one element return the NodeList

File: .gitignore [hunk 8]
```diff
@@ -9,6 +9,7 @@
 /bower_components
 
 # misc
+.DS_Store
```
Comment (rwjblue) [inline .gitignore]: I generally prefer to keep this as a system wide ignore

File: addon-test-support/helpers.js [hunk 9]
```diff
@@ -7,6 +7,13 @@ const DEFAULT_EVENT_OPTIONS = { canBubble: true, cancelable: true };
 const KEYBOARD_EVENT_TYPES = ['keydown', 'keypress', 'keyup'];
 const MOUSE_EVENT_TYPES = ['click', 'mousedown', 'mouseup', 'dblclick', 'mouseenter', 'mouseleave', 'mousemove', 'mouseout', 'mouseover'];
 
+// Reliable way to get the config?
```
Comment (rwjblue) [inline addon-test-support/helpers.js]: lets expose this as a function that users set in their `tests/test-helper.js`:

```js
// tests/test-helper.js or something
import config from '../config/environment';
import { setRootElementSelector } from 'ember-native-dom-helpers/test-support';

setRootElementSelector(config.APP.rootElement);
```

File: addon-test-support/helpers.js [hunk 10]
```diff
@@ -110,30 +117,100 @@ function buildKeyboardEvent(type, options = {}) {
...
+
+  @method first
+  @param {String} CSS selector to find one element in the test DOM
+  @param {HTMLElement} contextEl to query within, query from its contained DOM
+  @return {HTMLElement} first element found with matching selector
+  @public
+*/
+export function first(selector, contextEl) {
+  const el = find(selector, contextEl);
+  return (el instanceof NodeList) ? el[0] : el;
```
Comment (rwjblue) [inline addon-test-support/helpers.js]: Hmm, seems like this should be handled in `find`, no?  Meaning, instead of just returning a `NodeList` if passed in, we would return `selector[0]`...

File: package.json [hunk 11]
```diff
@@ -22,6 +22,7 @@
     "test": "ember try:each"
   },
   "dependencies": {
+    "broccoli-concat": "^3.2.2",
```
Comment (rwjblue) [inline package.json]: hmm, where is this used?

File: addon-test-support/helpers.js (hunk 10 above)
Comment (pixelhandler) [inline addon-test-support/helpers.js]: @rwjblue no the acceptance test helper `find` is mostly an alias to app.$(rootElement selector)

I reviewed `click` and `fillIn` helpers which use `el[0]` to only operate on the first item in a `find` result

File: package.json (hunk 11 above)
Comment (pixelhandler) [inline package.json]: In ember-cli I guess

File: addon-test-support/helpers.js (hunk 10 above)
Comment (pixelhandler) [inline addon-test-support/helpers.js]: @rwjblue `find` can return an HTMLElement or a NodeList (of elements). This is to simulate peoples use of `$('selelctor').doWhatever()` vs `$('selelctor')[0].doWhatever()`. The former affords operating on one or many items in the jQuery wrapped set. In testing we often use `find('selector')` vs `find('selector')[0]` to operate or compare an element.

File: package.json (hunk 11 above)
Comment (rwjblue) [inline package.json]: sorry, I think you misunderstood me in the issue.  You do not need to pin the version here like this, but you need to update your `yarn.lock` so that you can get the newer version. Simplest solution is to revert this change, and do `rm yarn.lock && yarn install` to update to the latest of all transitive deps.

File: .gitignore (hunk 8 above)
Comment (pixelhandler) [inline .gitignore]: Makes sense, will just use, https://help.github.com/articles/ignoring-files/#create-a-global-gitignore

File: addon-test-support/helpers.js (hunk 9 above)
Comment (pixelhandler) [inline addon-test-support/helpers.js]: @rwjblue is there a way to import the Resolver and lookup the config/environment module? We're not using a container here in the test helpers so can't just lookup config:environement.

Comment (rwjblue) [inline addon-test-support/helpers.js]: nope, which is why I suggested the helper function approach for now

Comment (pixelhandler) [inline addon-test-support/helpers.js]: @rwjblue seems odd to require that the consumer set the root element but maybe that can work as a blueprint to add to the consuming app tests/helpers directory. I can try that. I could live with using the `window.requirejs.entries` as well :) but sure I'll try it.

File: package.json (hunk 11 above)
Comment (pixelhandler) [inline package.json]: @rwjblue I did what you suggested and yarn added this line as well.

Comment (pixelhandler) [inline package.json]: @rwjblue when I remove that dependency I get the build error again:

```
Build failed.
The Broccoli Plugin: [SimpleConcatConcat: Concat: Vendor Styles/assets/vendor.css] failed with:
Error: Concat: Result is empty and allowNone is falsy.
```
